src/univariate/discrete/poissonbinomial.jl

```diff
@@ -24,35 +23,45 @@ External links:
 * [Poisson-binomial distribution on Wikipedia](http://en.wikipedia.org/wiki/Poisson_binomial_distribution)
 
 """
-struct PoissonBinomial{T<:Real} <: DiscreteUnivariateDistribution
-    p::Vector{T}
-    pmf::Vector{T}
-
-    function PoissonBinomial{T}(p::AbstractArray) where {T <: Real}
-        pb = poissonbinomial_pdf(p)
-        @assert isprobvec(pb)
-        new{T}(p, pb)
+mutable struct PoissonBinomial{T<:Real,P<:AbstractVector{T}} <: DiscreteUnivariateDistribution
+    p::P
+    pmf::Union{Nothing,Vector{T}} # lazy computation of the probability mass function
```

Comment:
Did you consider using an empty vector here? That would allow the struct to be immutable.

Reply:
Yes, I did. I wanted to try this approach first since arguably it is a bit "clearer" (`nothing` clearly indicates that it was not initialized whereas otherwise one has to check `isempty`) and the alternative would require to change `poissonbinomial_pdf` to an in-place function (which is clearly possible though). But I have no strong opinion here, I can compare it to the other approach.